Immutable iterator + iterate with token

oscartbeaumont opened the pull request:
This PR makes two changes:
 - Exposes the existing `Iter` and `IterPinRef` via the methods `StreamUnordered::iter` and `StreamUnordered::iter_pin`
 - Changes `Iter` and `IterPinRef`'s `Iterator::Item` type to return `(usize, _)`. The `usize` being the token of the stream.

Changing the `Item` type of an existing type would be a breaking change but as far as I can tell these types were incapable of being constructed as the fields are private and no methods I could see returned it constructed hence this change shouldn't be breaking.

I attempted an implementation for `StreamUnordered::iter_pin` but it's way outta my depth so this PR **is missing it's implementation.** I apologize for it not being functional but I really don't know what to do, any pointers would be appreciated and I can give it another shot.

## Use case

I am working on a project that queues many streams into a `StreamUnordered` that are related to a single connection (effectively a websocket). When this connection is closed I need to be able to clear all active streams so it would be useful if something like the following was possible:

```rust
let streams = StreamUnordered::new();

// Push some streams

// Remove all streams -> collected up the tokens to avoid having a mutable and immutable reference at the same time.
for id in streams.iter().map(|(token, stream)| token).collect::<Vec<_>>() {
    streams.remove(id);
}
```

Considerations:
 - It may make sense to expose the token in a similar fashion with the mutable iterators although this would be a breaking change. I am happy to update the PR if you have guidance on how to handle this.

This would also close #4.

jonhoo commented on src/lib.rs:
```diff
@@ -424,6 +424,34 @@ impl<S> StreamUnordered<S> {
+    }
+
+    /// Returns an immutable iterator that allows getting a reference to each stream in the set.
+    pub fn iter_pin(self: Pin<&Self>) -> IterPinRef<'_, S> {
```
I think we should just not expose this right now unless we have a concrete use-case that benefits from the pinning in the `&` case.

jonhoo commented on src/iter.rs:
```diff
@@ -97,13 +98,13 @@ impl<'a, S> Iterator for IterPinRef<'a, S> {
 }
 
-impl<S> ExactSizeIterator for IterPinRef<'_, S> {}
+impl<S: Unpin> ExactSizeIterator for IterPinRef<'_, S> {}
```
I don't think this extra bound should be necessary.

jonhoo commented on src/iter.rs:
```diff
@@ -88,7 +89,7 @@ impl<'a, S> Iterator for IterPinRef<'a, S> {
             self.task = next;
             self.len -= 1;
-            Some(Pin::new_unchecked(stream))
+            Some((task.id, Pin::new_unchecked(stream)))
```
I'm not 100% sure that it's safe for us to still access `task` here, so let's instead have:
```rust
let id = (*self.task).id;
```
directly next to `let stream` further up, and then just use `id` directly here.

jonhoo commented on src/lib.rs:
```diff
@@ -424,6 +424,34 @@ impl<S> StreamUnordered<S> {
     }
 
+    /// Returns an immutable iterator that allows getting a reference to each stream in the set.
+    pub fn iter(&self) -> Iter<'_, S>
```
I think it'll probably be pretty confusing for `iter` to include a token, but `iter_mut` not to. Let's instead add this as `iter_with_token` and add a `iter_mut_with_token` as the `&mut` version. Then in the next major version we can get rid of the non-token iterators and just have them all provide tokens instead.

oscartbeaumont commented on src/lib.rs:
```diff
@@ -443,9 +457,26 @@ impl<S> StreamUnordered<S> {
+
+    /// Returns an immutable iterator that allows getting a reference to each stream in the set.
+    pub fn iter_with_token(&self) -> IterWithToken<'_, S> {
+        let task = unsafe { *self.head_all.as_ptr() };
```
This function was implemented by me but derived from the impl of `StreamUnordered::iter_pin_mut_with_token`.

I have a feeling this first line is unsafe due to the fact that the method takes `&self` given the comments in the equivalent line within `iter_pin_mut_with_token`, but I am also very outta my depth in this level of unsafe so a full review of this function is in order.

All other functions remain identical minus some internal rearranging.

jonhoo replied:
Ah, so, for this you're going to want to use the `atomic_load_head_and_len_all` helper just below, which takes care of _exactly_ reading out the task pointer and length for you based on a `&self`. That should in turn let you make this function entirely safe :+1:

oscartbeaumont commented:
Thanks for your review!

I made the changes you requested and also changed:
 - Relaxed the unnecessary `Unpin` bounds.
 - The lint `broken_intra_doc_links` has been renamed `rustdoc::broken_intra_doc_links`
 - `use async_bincode::*;` in `tests/tcp.rs` test is causing compiler errors as `async_bincode::tokio` now exists and conflicts with the `tokio` crate in the scope of the file.
 - I mistakenly put `(usize, _)` instead of `(_, usize)` on one of the iterators
 - I added `iter_pin_mut_with_token` in extension to the methods you requested as this implementation is used for `iter_mut`, `iter_pin_mut` and `iter_mut_with_token` internally so it does need to exist in the crate.
 - I also attempted an implementation for `StreamUnordered::iter` which was previously unimplemented. **This needs a thorough review** as I am outta my depth in this level of unsafe.
 - I added some unit tests for the iterators

Ps. Really appreciate the content your making and was really cool to see you review this PR on stream!

jonhoo commented on src/iter.rs:
```diff
@@ -4,39 +4,46 @@ use core::marker::PhantomData;
 }
 
+/// Mutable iterator over all streams in the unordered set.
+///
+// Unfortunately, this ended up in the public API so remains to maintain backwards compatibility
```
That's totally fine, no need to call it out in the public docs :)
```suggestion
```

oscartbeaumont replied:
It shouldn't have been the case because of the `//` but removed it anyway.

jonhoo commented on src/iter.rs:
```diff
@@ -71,15 +92,31 @@ impl<'a, S: Unpin> Iterator for IterMut<'a, S> {
         }
         unsafe {
-            let stream = (*(*self.task).stream.get()).as_ref().unwrap();
+            let task = &*self.task;
+            let id = (*task).id;
+            let stream = (*task.stream.get()).as_ref().unwrap();
```
Let's do the same thing here as we changed with `iter_mut`, which is to _not_ construct a reference inside `*self.task`. That _might_ not be okay later in the same function. Instead, just repeat `*self.task` in the two places it's used.

jonhoo commented on tests/iter.rs:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    let result: (Vec<_>, Vec<_>) = streams.iter_with_token().unzip();
+    assert_eq!(result.1, vec![5, 4, 3, 2, 1]);
```
I don't think we want to guarantee anything about the _order_ in which `iter_with_token` yields the streams. Just call `sort` on the `Vec` before you compare it and you should be fine :+1:

jonhoo replied:
Same for the tests below.

jonhoo reviewed (changes requested):
This is looking quite good! Left a note about how you can avoid the unsafe, and then we're almost there :)

oscartbeaumont commented:
Ok, everything should be fixed up now!

jonhoo commented on src/lib.rs:
```diff
@@ -51,11 +50,12 @@ use core::sync::atomic::{AtomicBool, AtomicPtr};
 use futures_core::stream::{FusedStream, Stream};
 use futures_core::task::{Context, Poll};
 use futures_util::task::{ArcWake, AtomicWaker};
+use std::marker::PhantomData;
```
Ah, this needs to stay `core` for `no_std`-compatibility :+1:

oscartbeaumont replied:
Huh, I have no idea why this changed.

jonhoo commented on tests/iter.rs:
```diff
@@ -0,0 +1,45 @@
+use futures_util::stream;
+use streamunordered::StreamUnordered;
+
+// An `async move { i }` causes unpin issues with `iter_mut` and `iter_mut_with_token`.
```
You can just use [`std::future::ready`](https://doc.rust-lang.org/std/future/fn.ready.html) which also returns a future type that you can name and that is `Unpin` :+1:

jonhoo reviewed (approved):
Almost there!

oscartbeaumont commented:
Good catches! Should all be fixed now.

jonhoo reviewed (approved):
Excellent, thank you!

jonhoo commented:
Released in 0.5.3 :tada:
